core/node/crypto/client.go

```diff
@@ -6,136 +6,350 @@ import (
 	"math/big"
 
 	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
 	"github.com/ethereum/go-ethereum/common"
 	"github.com/ethereum/go-ethereum/core/types"
 	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/ethereum/go-ethereum/ethclient/simulated"
+	"github.com/prometheus/client_golang/prometheus"
 	"go.opentelemetry.io/otel/attribute"
 	"go.opentelemetry.io/otel/trace"
+
+	"github.com/river-build/river/core/node/infra"
 )
 
+type simulatedClientWrapper struct {
+	simulated.Client
+}
+
+var _ BlockchainClient = (*simulatedClientWrapper)(nil)
+
+// NewWrappedSimulatedClient returns a wrapped client that implements methods
+// missing on the simulated client which are occasionally used by abi-genned
+// contract code.
+func NewWrappedSimulatedClient(client simulated.Client) BlockchainClient {
+	return &simulatedClientWrapper{
+		Client: client,
+	}
+}
+
+func (scw *simulatedClientWrapper) CallContractAtHash(
+	ctx context.Context,
+	msg ethereum.CallMsg,
+	blockHash common.Hash,
+) ([]byte, error) {
+	bh, ok := scw.Client.(bind.BlockHashContractCaller)
+	if ok {
+		return bh.CallContractAtHash(ctx, msg, blockHash)
+	}
+
+	block, err := scw.BlockByHash(ctx, blockHash)
+	if err != nil {
+		return nil, err
+	}
+
+	return scw.CallContract(ctx, msg, block.Number())
+}
+
+func (scw *simulatedClientWrapper) CodeAtHash(
+	ctx context.Context,
+	account common.Address,
+	blockHash common.Hash,
+) ([]byte, error) {
+	bh, ok := scw.Client.(bind.BlockHashContractCaller)
+	if ok {
+		return bh.CodeAtHash(ctx, account, blockHash)
+	}
+
+	block, err := scw.BlockByHash(ctx, blockHash)
+	if err != nil {
+		return nil, err
+	}
+
+	return scw.CodeAt(ctx, account, block.Number())
+}
+
 type otelEthClient struct {
 	*ethclient.Client
-	tracer trace.Tracer
+	ethCalls *prometheus.CounterVec
+	tracer   trace.Tracer
 }
 
 var _ BlockchainClient = (*otelEthClient)(nil)
 
-// NewInstrumentedEthClient wraps an Ethereum client and adds open-telemetry tracing.
-func NewInstrumentedEthClient(client *ethclient.Client, tracer trace.Tracer) *otelEthClient {
-	return &otelEthClient{Client: client, tracer: tracer}
+// NewInstrumentedEthClient wraps an Ethereum client and adds open-telemetry tracing and metrics
+// collection.
+func NewInstrumentedEthClient(
+	client *ethclient.Client,
+	metrics infra.MetricsFactory,
+	tracer trace.Tracer,
+) *otelEthClient {
+	var ethCalls *prometheus.CounterVec
+	if metrics != nil {
+		ethCalls = metrics.NewCounterVecEx(
+			"eth_calls",
+			"Number of eth_calls made by an instrumented client",
+			"method_name",
```

Comment:
please add `chain_id` as first label.

please take a look into adding `status` as last label.
`ok` on success
revert reason if it's possible to extract it like this:
https://github.com/river-build/river/blob/b22df64b5a4e86ce78e81f3298704f85f938ac4e/core/node/base/error.go#L267
`revert` if error is rpc.DataError but reason can't be extracted
`fail` if this is some other error.

Reply:
See updated output above for examples of capturing revert reasons